### PR TITLE
Make it possible to donate to plugin creators from within piwik

### PR DESCRIPTION
**Pull request by sgiehl**
As discussed at the meetup in Berlin we should make it possible to donate to plugin creators directly from within Piwik.

I've updated the list of plugins (within the piwik admin) to show an additional text if there are donation possibilities available:

image

Clicking on the link opens a small popover, showing all available possibilities.

image

Any suggestions / thoughts for improvements?
Guess the displayed text could be improved, maybe someone has a good suggestion for a better wording?!

### CODE REVIEW

**Comment by tsteur** [comment]
I haven't tested it, only had a quick look at the code and looks good. Also like the idea of displaying it in the app

File: plugins/CorePluginsAdmin/templates/macros.twig [hunk 1]
```diff
@@ -187,6 +187,29 @@
...
+                                    <div id="overlay-{{ name|escape('html_attr') }}" class="donation-overlay ui-confirm" title="{{ 'CorePluginsAdmin_LikeThisPlugin'|translate }}">
+                                        <p>{{ 'CorePluginsAdmin_CommunityContributedPlugin'|translate }}</p>
+                                        <p>{{ 'CorePluginsAdmin_ConsiderDonatingCreatorOf'|translate("<b>" ~ name ~ "</b>")|raw }}</p>
+                                        <div class="donation-links">
+                                        {% if plugin.info.donate.paypal is defined and plugin.info.donate.paypal %}
+                                            <a class="donation-link paypal" target="_blank" href="https://www.paypal.com/cgi-bin/webscr?cmd=_donations&item_name=Piwik%20Plugin%20{{ name|escape('url') }}&bn=PP%2dDonationsBF%3abtn_donateCC_LG%2egif%3aNonHosted&business={{ plugin.info.donate.paypal|escape('url') }}"><img src="https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif" height="30"/></a>
```
Comment (mattab) [inline plugins/CorePluginsAdmin/templates/macros.twig]: for privacy, we should not load resources from outside of Piwik - can you bundle the paypal image in Piwik?

File: plugins/CorePluginsAdmin/templates/macros.twig [hunk 2]
```diff
@@ -187,6 +187,29 @@
...
+                                        {% endif %}
+                                        {% if plugin.info.donate.flattr is defined and plugin.info.donate.flattr %}
+                                            <a class="donation-link flattr" target="_blank" href="{{ plugin.info.donate.flattr }}"><img class="alignnone" title="Flattr" alt="" src="https://api.flattr.com/button/flattr-badge-large.png" height="29" /></a>
+                                        {% endif %}
+                                        {% if plugin.info.donate.bitcoin is defined and plugin.info.donate.bitcoin %}
+                                            <div class="donation-link bitcoin"><a class="coinbase-button" data-height="30" data-button-style="donation_small" data-code="{{ plugin.info.donate.bitcoin|escape('html_attr') }}" href="https://coinbase.com/checkouts/{{ plugin.info.donate.bitcoin|escape('url') }}" target="_blank">Donate With Bitcoin</a></div>
+                                            <script src="https://coinbase.com/assets/button.js" type="text/javascript"></script>
```
Comment (mattab) [inline plugins/CorePluginsAdmin/templates/macros.twig]: let's not load external JS in Piwik (for security) - could we link to the page to donate via bitcoins?

**Comment by mattab** [comment]
Great idea @sgiehl - I put couple comments, otherwise good to merge

File: plugins/CorePluginsAdmin/templates/macros.twig (hunk 1 above)
Comment (sgiehl) [inline plugins/CorePluginsAdmin/templates/macros.twig]: Ok. Will move that images to piwik

File: plugins/CorePluginsAdmin/templates/macros.twig (hunk 2 above)
Comment (sgiehl) [inline plugins/CorePluginsAdmin/templates/macros.twig]: To be honestly I don't how to solve that. I'm not that familiar with bitcoins, and I don't know any services providing that functionality. Or is it common just to place a ```bitcoin:XXXXXXXXXX``` link? Maybe @MagicFab has a suggestion?

**Comment by mattab** [comment]
Re; bitcoin, alternatively you could just show the bitcoin address and users could copy paste it to send money to?

**Comment by sgiehl** [comment]
@mattab I've changed the bitcoin stuff. I've updated the screenshot in the PR description above.
Do you have suggestions to improve the text? Otherwise guess we could merge!

**Comment by MagicFab** [comment]
Bitcoin could be specified in one of two ways:

- a URL to a bitcoin processor/landing page ( two examples: https://bitpay.com/775767/donate -> Bitcoin processor page, http://legoutdulibre.com/bitcoin -> custom landing page)
- Bitcoin URI ( bitcoin:1xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx )

It is against best practices to reuse a single address but possible (and may be used for any of many valid reasons). Bitcoin payment processors take care of generating a new address, forwarding/converting payments, etc.

When clicking a Bitcoin URI it will open the local Bitcoin app for payment if one is installed (for example Ethereum). 

I can test this if/when implemented.

**Comment by sgiehl** [comment]
@MagicFab Plugin developers can set exactly one bitcoin address per plugin, so we need to use that one.
I don't have an account at bitpay.com, but I guess it's mostly like coinbase.com. Problem with that is, we do not want to include external javascripts. So I guess using the bitcoin uri might be the best solution for now.

**Comment by MagicFab** [comment]
@sgiehl ok, using a single Bitcoin address is fine.

Adding the possibility to have a direct URL link to self-shoted pages or to service like Bitpay does not require Javascript, however. It's a direct link, just like the current Paypal / Flatter links.

**Comment by sgiehl** [comment]
@MagicFab not sure if I got you wrong, that would require to let the user define the full URL to their bitcoin donation page. Guess it should be possible to check if the value is a full url or a bitcoin address and handle the link according to that.

@mattab what do you think about that?

**Comment by mattab** [comment]
@sgiehl sounds good to me if `bitcoin` can be either a bitcoin address, or a URL to a custom bitcoin donate page. btw let's also mention this possibility in the dev guide in http://developer.piwik.org/guides/distributing-your-plugin#prepare-your-plugin

**Comment by sgiehl** [comment]
@mattab ok. do we have any validation checks on the marketplace for the value of the donation fields?

**Comment by mattab** [comment]
Good point! I just found this code:
```
            if (empty($donate['bitcoin'])
                || !ctype_xdigit($donate['bitcoin'])
                || strlen($donate['bitcoin']) < 10
                || strlen($donate['bitcoin']) > 100) {
                $donate['bitcoin'] = null;
            }
```

so  URLs will not work in the `bitcoin` field! 

Could you get in touch with Thomas to have this changed?

**Comment by sgiehl** [comment]
Hm... on the other hand I'm not sure if it would really be good to link to a custom URL. That would allow users to inject "bad" links in piwik and the marketplace. Maybe it would be better to whitelist some urls (domains), but I don't know which services might be good to whitelist

**Comment by sgiehl** [comment]
@mattab: I'd suggest to leave it as it is for now. Maintaining kind of a whitelist for serious bitcoin donation service does not really seem to be a good option.. So guess we can merge, if the wording is fine for you.

**Comment by mattab** [comment]
Nice change!

**Pull request merged**